### PR TITLE
SONARJAVA-1270 RSPEC-3318 Data should not be stored in session

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/DataStoredInSessionCheck.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.SyntaxToken;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+
+import java.util.List;
+
+@Rule(
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/DataStoredInSessionCheck.java]: MINOR Reorder annotation arguments to match the order of declaration. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS3340)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: java-checks/src/main/java/org/sonar/java/checks/DataStoredInSessionCheck.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+  priority = Priority.CRITICAL,
+  tags = {"cwe", "security"})
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.INPUT_VALIDATION_AND_REPRESENTATION)
+@SqaleConstantRemediation("20min")
+public class DataStoredInSessionCheck extends AbstractMethodDetection {
+
+  private static final MethodInvocationMatcherCollection REQUEST_OR_COOKIE_DATA_RETRIEVAL = MethodInvocationMatcherCollection.create(
+    MethodMatcher.create().typeDefinition("javax.servlet.http.Cookie").name(NameCriteria.startsWith("get")).withNoParameterConstraint(),
+    MethodMatcher.create().callSite(TypeCriteria.is("javax.servlet.http.HttpServletRequest")).name(NameCriteria.startsWith("get")).withNoParameterConstraint(),
+    MethodMatcher.create().callSite(TypeCriteria.is("javax.servlet.http.HttpServletRequest")).name(NameCriteria.startsWith("get")).addParameter(TypeCriteria.anyType()));
```
Comment (mpaladin) [inline java-checks/src/main/java/org/sonar/java/checks/DataStoredInSessionCheck.java]: @m-g-sonar this MethodMatcher is not needed, the previous line with `withNoParameterConstraint` covers this one.

File: java-checks/src/main/java/org/sonar/java/checks/DataStoredInSessionCheck.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+  @Override
+  protected List<MethodMatcher> getMethodInvocationMatchers() {
+    return ImmutableList.of(MethodMatcher.create().typeDefinition("javax.servlet.http.HttpSession").name("setAttribute").withNoParameterConstraint());
```
Comment (mpaladin) [inline java-checks/src/main/java/org/sonar/java/checks/DataStoredInSessionCheck.java]: @m-g-sonar you can add parameters constraint as you require to have two parameters in line 66.

File: java-checks/src/main/java/org/sonar/java/checks/DataStoredInSessionCheck.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    if (expressionToEvaluate.is(Tree.Kind.IDENTIFIER)) {
+      Symbol variable = ((IdentifierTree) expressionToEvaluate).symbol();
+      ExpressionTree lastAssignmentOrDeclaration = ReassignmentFinder.getClosestReassignmentOrDeclarationExpression(mit, variable);
+      if (lastAssignmentOrDeclaration != null && !usedBetween(variable, lastAssignmentOrDeclaration, mit)) {
+        expressionToEvaluate = lastAssignmentOrDeclaration;
+      }
+    }
+
+    if (isRequestOrCookieDataRetrieval(expressionToEvaluate)) {
+      addIssue(mit, "Make sure the user is authenticated before this data is stored in the session.");
```
Comment (mpaladin) [inline java-checks/src/main/java/org/sonar/java/checks/DataStoredInSessionCheck.java]: @m-g-sonar you can try the new api: `reportIssue(mit.methodSelect(), "Make sure the user is authenticated before this data is stored in the session.");`

**Pull request merged**